### ISSUE TITLE
Should (internal) platform libraries be able to ignore class modifiers in other platform libraries.

### ISSUE THREAD

**Issue by lrhn** [dart-lang/language#2833]
We want to declare `bool` as `final class bool { ... }` to match the declaration to the current behavior of not being usable as superclass, interface or mixin.

However, our internal implementations seem to rely on the ability to ignore that constraint, and not just inside the same library.

For example, the web code has an `interceptors.dart` *libraray* which does `class JSBool ... implements bool {...}`.

When enabling the feature for platform libraries, and adding `final` to the `bool` class, that interceptor class stops compiling.

So, short term and long term, how should we fix that:

* Short term, we can allow platform libraries to ignore `final`/`base`/`interface` modifiers in other platform libraries. Should we? It's not a feature we provide to other people, so why do we need it? _(Does it suggest that `final`/`base`/`interface`, but not `sealed`, should be ignorable in the same package, instead of just same library? )_

* Long term, should we restructure our backend libraries so that the `JSBool` class is declared in the same library as `bool`, possibly using a platform-specific patch/augmentation file? (But without exposing the type publicly, so possibly have a `dart:_core` library with both, where `dart:core` exports only the public-facing types. We could combine all our platform libraries into a few big `dart:_impl` libraries, one per platform combination, and have `dart:core`, `dart:async` etc. export from there.)

WDYT?

@leafpetersen @munificent @kallentu @eernstg @stereotype441 @natebosch @jakemac53 
@rakudrama @mkustermann @annagrin

**Comment by eernstg** [comment]
_tl;dr Allow platform libraries to ignore class modifiers, but don't try to reorganize the code rd;lt_

> It's not a feature we provide to other people, so why do we need it?

Many platform library defined classes have built-in semantics, that is, their declarations already mean something which is not expressible in Dart, and possibly very different from the semantics of that same syntax, had it not been treated in a special way by the analyzer and compilers. Allowing certain class modifiers to be treated in a special way (e.g.: to be ignored) would simply be yet another case where such things occur.

It would be possible to allow arbitrary violations of the new constraints associated with class modifiers for now, and then the enforcement could be made more strict in the future. If `JSBool` is the only problem then that particular class could be the only thing that gets this special treatment.

> short term and long term ...

We have had so many situations where the need for special treatment of packages (or some other entity bigger than a library) has come up. Let's call it a _module_, which could be a package.

The conflict that always comes up right after is that of separate compilation: If we allow code generation to depend on an entire module then that module is essentially a single library anyway because everything in there needs to be compiled at the same time.

Still, if that module contains several libraries then we could take the opportunity to provide the features offered by that module in terms of a _module interface_, which would be a mechanism that could be used to export some declarations in some libraries, and not others.

In this case I think there is need for a proper language mechanism, rather than trying to solve a source code organization problem in platform libraries which is actually relevant to the language and community as a whole.

**Comment by lrhn** [comment]
>  If `JSBool` is the only problem then that particular class ...

It's not. 
See [log](https://logs.chromium.org/logs/dart/buildbucket/cr-buildbucket/8789637527993816529/+/u/build_linux-ia32_for_benchmarking/stdout) of [this CL](https://dart-review.googlesource.com/c/sdk/+/281683). It does seem to be entirely web libraries which have allowed themselves to ignore the existing restrictions.

**Comment by jakemac53** [comment]
> * (Does it suggest that `final`/`base`/`interface`, but not `sealed`, should be ignorable in the same package, instead of just same library? )

We would need to ensure this doesn't mess with exhaustiveness checking, especially when doing modular compilation which can be at a finer granularity than packages.

**Comment by leafpetersen** [comment]
Is this something we're still considering?

**Comment by Wdestroier** [comment]
Hopefully augmentations can solve the problem. If the language needs to bypass its own rules to interop with its own sdk code, then allowing app developers to bypass restrictions imposed by package authors would be fair too.

**Comment by lrhn** [comment]
The current implementation has a list of libraries which can ignore privacy. Those libraries are now also allowed to ignore class modifiers just like pre-feature libraries. (So not `sealed` - nobody ignores `sealed`!)

We can always twiddle with our internals, but for now it works.
